components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java

```diff
@@ -261,16 +268,43 @@ private void validateRequestObject(OAuthAuthzRequest oAuthAuthzRequest) throws P
                 parameters.setResponseType(oAuthAuthzRequest.getResponseType());
                 parameters.setTenantDomain(getSPTenantDomainFromClientId(oAuthAuthzRequest.getClientId()));
 
-                RequestObject requestObject = OIDCRequestObjectUtil.buildRequestObject(oAuthAuthzRequest, parameters);
+                requestObject = OIDCRequestObjectUtil.buildRequestObject(oAuthAuthzRequest, parameters);
                 if (requestObject == null) {
                     throw new ParClientException(OAuth2ErrorCodes.INVALID_REQUEST, ParConstants.INVALID_REQUEST_OBJECT);
                 }
             }
+            return requestObject;
         } catch (RequestObjectException e) {
             if (OAuth2ErrorCodes.SERVER_ERROR.equals(e.getErrorCode())) {
                 throw new ParCoreException(e.getErrorCode(), e.getMessage(), e);
             }
             throw new ParClientException(e.getErrorCode(), e.getMessage(), e);
         }
     }
+
+    /**
+     * Validate PKCE parameters for PAR requests.
+     * According to FAPI(5.2.2-18), PAR requests require to use PKCE (RFC7636) with S256 as the code challenge method.
+     * <a href="https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server">...</a>
+     *
+     * @param paramMap parameter map
+     * @throws ParClientException if PKCE validation fails
+     */
+    private void validatePKCEParameters(Map<String, String> paramMap) throws ParClientException {
+
+        String codeChallenge = paramMap.get(OAuthConstants.OAUTH_PKCE_CODE_CHALLENGE);
+        String codeChallengeMethod = paramMap.get(OAuthConstants.OAUTH_PKCE_CODE_CHALLENGE_METHOD);
+
+        if (StringUtils.isEmpty(codeChallenge)) {
+            throw new ParClientException(OAuth2ErrorCodes.INVALID_REQUEST,
+                    "Mandatory parameter code_challenge, not found in the request.");
+        }
+        if (StringUtils.isEmpty(codeChallengeMethod)) {
```

Comment:
it looks like what the spec is actually mandating is the use of S256 code challenge method. if the code challenge method is not sent, identity server defaults to S256. So do we need to have a mandatory code challenge method check? wouldn't it be sufficient to check if the code challenge method is sent then that is S256.

Reply:
Spec mandates the **use of PKCE** and **having the code challenge method S256** with **PAR** auth requests. Here we are validating whether PKCE parameters are sent in the PAR auth request even if PKCE is not mandated at the app level. Code challenge method S256 validation was taken to this validation as well according to the review comment received during the code review to have FAPI validations at one place without depending on the existing validations.

Reply:
Logically the reason to require the challenge method is to make sure that S256 is used. In identity server if the method is not sent it defaults to S256. 
But in the relevant OAuth [spec](https://datatracker.ietf.org/doc/html/rfc7636#section-4.3) it mentions defaulting to 'plain', therefor +1 to keeping the check as it is.